src/xmipp/libraries/reconstruction/tomo_filter_coordinates.cpp

```diff
@@ -0,0 +1,232 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+ * 02111-1307  USA
+ *
+ *  All comments concerning this program package may be sent to the
+ *  e-mail address 'xmipp@cnb.csic.es'
+ ***************************************************************************/
+
+#include "tomo_filter_coordinates.h"
+#include <core/metadata_extension.h>
+#include <core/metadata_vec.h>
+#include <iostream>
+#include <core/utils/time_utils.h>
+
+
+// --------------------------- INFO functions ----------------------------
+
+void ProgTomoFilterCoordinates::readParams()
+{
+	fnInTomo = getParam("--inTomo");
+	fnMask = getParam("--mask");
+	fnInCoord = getParam("--coordinates");
+	radius = getIntParam("--radius");
+   	fnOut = getParam("-o");
+}
+
+
+void ProgTomoFilterCoordinates::defineParams()
+{
+	addUsageLine("This program generates a new metatada coordinates file with local information given by the input tomogram (e.g. local ");
+	addUsageLine("resolution map). \n"); 
+	addUsageLine("Given a set of coordinates and a tomogram, local statistics will be calculated for each coordinate (given by the radius). "); 
+	addUsageLine("Optionally a mask can be provided, and then those coordinates outside of the mask (value = 0) will be removed.");
+	
+	addParamsLine("  --inTomo <mrcs_file=\"\">                     : Input volume (mask or resolution map).");
+	addParamsLine("  [--mask <xmd_file=\"\">]                      : Input xmd file containing the 3D coordinates.");
+	addParamsLine("  --coordinates <xmd_file=\"\">                 : Percentile resolution threshold.");
+	addParamsLine("  --radius <radius=50>                          : Radius of the neighbourhood of the coordinates to get resolution score.");
+	addParamsLine("  -o <outCoord=\"filteredCoordinates3D.xmd\">   : Output file containing the filtered 3D coordinates.");
+}
+
+
+// --------------------------- HEAD functions ----------------------------
+
+void ProgTomoFilterCoordinates::filterCoordinatesWithMask(MultidimArray<int> &inputVolume)
+{
+	Point3D<int> coord3D;
+	for (size_t i = 0; i < inputCoords.size(); i++)
+	{
+		coord3D = inputCoords[i];
+
+		if(coord3D.z < 0 || coord3D.z > (zDim-1) || coord3D.y < 0 || coord3D.y > (yDim-1) || coord3D.x < 0 || coord3D.x > (xDim-1))
+		{
+			std::cout << "WARNNING: Coordinate at (x=" << coord3D.x<< ", y=" << coord3D.y << ", z=" << coord3D.z << ") erased due to its out of the mask." << std::endl;
+			inputCoords.erase(inputCoords.begin()+i);
+			i--;
```

Comment:
this will break if you remove the first element in the list (size_t underflow).
Consider rewriting the whole method using `std::remove_if` https://en.cppreference.com/w/cpp/algorithm/remove

Reply:
We replace datatype from size_t to int to avoid the error. the line i-- could result in a negative element that is not supported by size_t, the int solves this problem and the i++ ensure that the index will be greater than 0.

Reply:
> the int solves this problem and the i++ ensure that the index will be greater than 0

But is not implemented.

Reply:
> the int solves this problem and the i++ ensure that the index will be greater than 0

But is not implemented.

Reply:
I think is something related to the PR viewer. In my local code it is fixed and up to date with the remote branch. Also if you visualize the changes in commit 7530da9, you can see it is changed to int datatype (also if you explore the code in the repo the file shows the changes).

Reply:
It is now implemented :) size_t has been substituted by int.